lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,662 @@
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
+            }
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = $this->skipNodeAndChildren($node);
+            } else {
+                $node = $this->nextNode($node);
+            }
+        }
+
+        // Optimize data-hero element if defined.
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Fall back to auto-detected hero image if available.
+        if ($heroImageCandidate) {
+            return [$heroImageCandidate];
+        }
+
+        // No hero images to optimize.
+        return [];
+    }
+
+    /**
+     * Detect a hero image with the data-hero attribute.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image, or null if none detected.
+     */
+    private function detectImageWithDataHero(DOMElement $element)
+    {
+        if (!$element->hasAttribute(Attribute::DATA_HERO)) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
+            return new HeroImage(
+                $element->getAttribute(Attribute::SRC),
+                $element->getAttribute(Attribute::MEDIA),
+                $element->getAttribute(Attribute::SRCSET),
+                $element
+            );
+        }
+
+        if ($this->isAmpEmbed($element)) {
+            return $this->getPlaceholderImage($element);
+        }
+
+        $cssBackgroundImage = $this->getCssBackgroundImageUrl($element);
+
+        if (Url::isValidImageSrc($cssBackgroundImage)) {
+            return new HeroImage(
+                $cssBackgroundImage,
+                $element->getAttribute(Attribute::MEDIA),
+                $element->getAttribute(Attribute::SRCSET),
+                $element
+            );
+        }
+
+        return null;
+    }
+
+    /**
+     * Detect a hero image candidate.
+     *
+     * The hero image here can come from one of <amp-img>, <amp-video>, <amp-iframe>, <amp-video-iframe>.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidate(DOMElement $element)
+    {
+        if (
+            $element->hasAttribute(Attribute::LAYOUT)
+            && $element->getAttribute(Attribute::LAYOUT) === Layout::NODISPLAY
+        ) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
```

Comment:
To anticipate `img` being allowed in AMP per https://github.com/ampproject/amphtml/issues/30442, we could include `img` here:

```suggestion
        if ($element->tagName === Extension::IMAGE || $element->tagName === Tag::IMG) {
```